### ISSUE TITLE
L1 penalty set small weights to 0

### ISSUE THREAD

**Issue by coco90417** [keras-team/keras#18388]
Dear team,

I have noticed that L1 penalty does not strictly force weights to 0. I wonder what is your view on below tentative approach that forces small weights to strictly 0?

class SmallWeightsToZero(Constraint):
    def __call__(self, w):
        mask = tf.abs(w) < tf.keras.backend.epsilon()  
        w = tf.where(mask, tf.zeros_like(w), w)
        return w 

model.add(layers.Dense(units = 1, 
                          kernel_regularizer = regularizers.L1(l1=l1),
                          kernel_constraint = SmallWeightsToZero()))

**Comment by tilakrayal** [comment]
@coco90417,
Unfortunately with the code provided above, we are unable to debug the issue. Could you please provide the complete standalone code and the tensorflow version you are using. Thank you!

**Comment by coco90417** [comment]
Thank you for looking into this.

Below is what you could reproduce the issue. 

import tensorflow as tf
import numpy as np
from tensorflow import keras

## 1000 observations; 20 predictors, 1 of predictors is redundant of all 0s
## so you would be expecting weight for the first predictor to be 0 with L1 penalty

x_training = tf.concat(
    [
        tf.zeros([1000, 1]),
        tf.random.normal([1000, 19], seed = 1),
    ], axis = 1)

y_training = tf.random.normal([1000], seed = 1)

from keras import models, layers, regularizers, optimizers

# lasso regression with tensorflow
model = models.Sequential()
model.add(layers.InputLayer(input_shape = (20, )))
model.add(layers.Dense(units = 1, 
                       kernel_regularizer = regularizers.L1(l1=1000)))
model.compile(optimizer = optimizers.Adam(), 
                  loss = "mse") 

fit = model.fit(x_training, y_training, epochs = 1000)

model.get_weights()

[array([[**-1.15685923e-06**],
        [ 5.27249262e-05],
        [-1.37503495e-04],
        [-1.28477444e-04],
        [-2.45119820e-06],
        [ 1.45760217e-04],
        [ 1.54898667e-04],
        [ 2.10382195e-05],
        [ 1.31973664e-06],
        [-1.33556408e-04],
        [ 4.12663186e-06],
        [ 1.44984601e-05],
        [-1.64269970e-04],
        [-1.78692422e-04],
        [-1.08664912e-04],
        [ 2.58929811e-04],
        [-5.15903737e-06],
        [ 2.46357895e-05],
        [ 7.39442373e-06],
        [-1.24024616e-04]]),
 array([0.00742478])]

The weight is not even small for predictor 1. The weight constraint was what I thought might address it, but would to hear about your recommended approach.

**Comment by coco90417** [comment]
> @coco90417, Unfortunately with the code provided above, we are unable to debug the issue. Could you please provide the complete standalone code and the tensorflow version you are using. Thank you!

Please see my response above. Much appreciated.

**Comment by coco90417** [comment]
@tilakrayal Let me know how may I help on this issue.
